gpcontrib/arenadata_toolkit/arenadata_toolkit.c

```diff
@@ -1,4 +1,169 @@
+	TupleDesc tupdesc;
+} user_fctx_data;
+
+/*
+ * Name of file must be "XXX.X" or "XXX"
+ * where X is digit. Count of digits is not set.
+ */
+static bool should_skip_file(const char *filename)
+{
+	int pos = 0;
+	int filenamelen = strlen(filename);
+	bool found_dot = false;
+
+	for (; pos < filenamelen; pos++)
+	{
+		if (isdigit(filename[pos]))
+			continue;
+
+		if ('.' == filename[pos] && !found_dot && pos != 0)
+		{
+			found_dot = true;
+			continue;
+		}
+
+		return true;
+	}
+
+	return false;
+}
+
+PG_FUNCTION_INFO_V1(adb_get_relfilenodes);
+Datum adb_get_relfilenodes(PG_FUNCTION_ARGS)
+{
+	Oid              datoid = MyDatabaseId;
+	Oid              tablespace_oid = PG_GETARG_OID(0);
+
+	struct dirent   *direntry;
+	user_fctx_data  *fctx_data;
+	FuncCallContext *funcctx;
+
+	if (tablespace_oid == GLOBALTABLESPACE_OID)
+		datoid = 0;
+
+	if (SRF_IS_FIRSTCALL())
+	{
+		MemoryContext oldcontext;
+
+		funcctx = SRF_FIRSTCALL_INIT();
+		oldcontext = MemoryContextSwitchTo(funcctx->multi_call_memory_ctx);
+
+		fctx_data = palloc(sizeof(user_fctx_data));
+		fctx_data->datpath = GetDatabasePath(datoid, tablespace_oid);
+		fctx_data->dirdesc = AllocateDir(fctx_data->datpath);
+
+		if (!fctx_data->dirdesc)
+		{
+			/* Nothing to do: empty tablespace (maybe it has been just created)*/
+			MemoryContextSwitchTo(oldcontext);
+			SRF_RETURN_DONE(funcctx);
+		}
+
+		if (get_call_result_type(fcinfo, NULL, &fctx_data->tupdesc)
+				!= TYPEFUNC_COMPOSITE)
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("return type must be a row type")));
+
+		funcctx->attinmeta = TupleDescGetAttInMetadata(fctx_data->tupdesc);
+		funcctx->user_fctx = fctx_data;
+
+		MemoryContextSwitchTo(oldcontext);
+	}
+
+	funcctx = SRF_PERCALL_SETUP();
+	fctx_data = (user_fctx_data *) funcctx->user_fctx;
+
+	while ((direntry = ReadDir(fctx_data->dirdesc, fctx_data->datpath)) != NULL)
+	{
+		struct stat fst;
+		Datum       values[10];
+		bool        nulls[10];
+		char       *relfilenode, *ptr, *filename;
+		Oid         reloid;
+		Oid         relfilenode_oid;
+		HeapTuple   tuple;
+
+		CHECK_FOR_INTERRUPTS();
+
+		if (direntry->d_type == DT_DIR)
+			continue;
+
+		if (should_skip_file(direntry->d_name))
+			continue;
+
+		filename = psprintf("%s/%s", fctx_data->datpath, direntry->d_name);
+
+		if (stat(filename, &fst) < 0)
+		{
+			if (errno == ENOENT)
+				continue;
+
+			ereport(ERROR,
+					(errcode_for_file_access(),
+					 errmsg("could not stat file \"%s\": %m",
+							filename)));
+		}
+
+		memset(values, 0, sizeof(values));
+		memset(nulls, 0, sizeof(nulls));
+
+		values[0] = Int16GetDatum(GpIdentity.segindex);
+		values[1] = Int16GetDatum(GpIdentity.dbid);
+		values[2] = DatumGetObjectId(datoid);
+		values[3] = DatumGetObjectId(tablespace_oid);
+		values[4] = CStringGetTextDatum(filename);
+
+		relfilenode = pstrdup(direntry->d_name);
+		ptr = strchr(relfilenode, '.');
+		if (ptr)
+			*ptr = '\0';
+
+		relfilenode_oid = DirectFunctionCall1(oidin,
```

Comment:
The `oidin` function (more precisely, its internal implementation `oidin_subr`) can produce a bunch of errors `ereport(ERROR...`
Maybe we should ignore them, instead of stopping with an error?